### PR TITLE
Add features to rknn detector

### PR DESCRIPTION
**Pull request by MarcA711**
Added features:
- Add checks to make sure that the user sets the correct options in his config (for example `input_pixel_format: bgr` instead of rgb)
- Let the user choose the yolov8 model: 
  - yolov8n is still included in the image, but now it is downloaded during the image build process and not added to the frigate repository and copied over.
  - The user can choose another model (yolov8{s,m,l,x}); for this the user needs to mount a directory to his system and the model will be downloaded to that directory if it does not exist.
  - The user can also use a custom model that he provides using the `model: path:` option; this overwrites the provided yolov8 models
- adds the ability to set the core mask of the npu
- adds options for postprocessing, but I have some questions here @NickM-27 :
  - min_score: the minimum score for a possible detection to be considered a detection; this option is basically the same as `object: min_score:`. It would be better, if I could access this option directly, but I don't know how.
  - nms_thresh: IoU threshold used in nms algorithm; this basically controls whether two detection boxes that overlap are actually one object or not. This option also fits better in the `objects` block of the config, rather than the `detectors` block. However, it would be confusing for other users to add an option that only some detectors actually use. So, if this option gets added to the object block, it should be implemented in all other detectors. However, then we should also add all parameters used in postprocessing of other models. Therefore, I think it is best to keep it in the rknn detector block for now. But in the future, we could maybe implement a common `postprocessing.py` file that all detectors can use to do their postprocessing and that exposes the most important parameters to the config.
- Updated docs explaining the new configuration options.

### CODE REVIEW

**Comment by NickM-27** [comment]
I don't think there is a reason to have a separate config for min_score, I would suggest just following the other detectors that use 0.4 as the min score. Also, frigate already does an NMSboxes post processing in video.py, I think it is inefficient to do this twice

File: docs/docs/configuration/object_detectors.md [hunk 1]
```diff
@@ -309,14 +309,23 @@ RKNN support is provided using the `-rk` suffix for the docker image. Moreover,
...
 This `config.yml` shows all relevant options to configure the detector and explains them. All values shown are the default values (except for one). Lines that are required at least to use the detector are labeled as required, all other lines are optional.
+
 ```yaml
 detectors:                            # required
   rknn:                               # required
     type: rknn                        # required
+    # core mask for npu
+    core_mask: 0
+    # yolov8 model in rknn format to use; allowed calues: n, s, m, l, x
+    yolov8_rknn_model: n
```
Comment (NickM-27) [inline docs/docs/configuration/object_detectors.md:320]: maybe we can map these to be more user friendly? for example

```
{
"tiny": "n",
"small": "s",
...
}
```

File: docs/docs/configuration/object_detectors.md [hunk 2]
```diff
@@ -309,14 +309,23 @@ RKNN support is provided using the `-rk` suffix for the docker image. Moreover,
...
+    # minimal confidence for detection
+    min_score: 0.5
```
Comment (NickM-27) [inline docs/docs/configuration/object_detectors.md:322]: I don't think this should be configurable, it is not configurable for other models

File: docs/docs/configuration/object_detectors.md [hunk 3]
```diff
@@ -309,14 +309,23 @@ RKNN support is provided using the `-rk` suffix for the docker image. Moreover,
...
+    # determines whether two overlapping boxes should be combined
+    nms_thresh: 0.45
```
Comment (NickM-27) [inline docs/docs/configuration/object_detectors.md:324]: I don't think this should be configurable, it is not configurable for other models

File: docs/docs/configuration/object_detectors.md [hunk 4]
```diff
@@ -309,14 +309,23 @@ RKNN support is provided using the `-rk` suffix for the docker image. Moreover,
...
 
 model:                                # required
   # path to .rknn model file
-  path: /models/yolov8n-320x320.rknn
+  path:
```
Comment (NickM-27) [inline docs/docs/configuration/object_detectors.md:328]: there is still a default here right?

File: docs/docs/configuration/object_detectors.md [hunk 5]
```diff
@@ -325,4 +334,60 @@ model:                                # required
...
+```bash
+$ cat /sys/kernel/debug/rknpu/load
+>> NPU load:  Core0:  0%, Core1:  0%, Core2:  0%,
+```
+
+:::
+
+- By default the rknn detector uses the yolov8**n** model (`yolov8_rknn_model: n`). This model comes with the image, so no further steps than those mentioned above are necessary.
+- If you want to use are more precise model, you can set `yolov8_rknn_model:` to `s`, `m`, `l` or `x`. But additional steps are required:
+    1. Mount the directory `/model/download/` to your system using one of the below methods. Of course, you can change to destination folder.
```
Comment (NickM-27) [inline docs/docs/configuration/object_detectors.md:376]: we should recommend users to use `/config/model_cache/rknn/` similar to how other detectors already do this

**Review by NickM-27 — CHANGES_REQUESTED** [review]

**Comment by MarcA711** [comment]
Alright, I didn't know that frigate already does postprocessing. If I got it right, I only need to remove all detections with score smaller than 0.4 and return the 20 (or less) detections with the highest score. No nms or other postprocessing?

**Comment by NickM-27** [comment]
Correct, you can see these links below:

edgetpu detector that limits based on score of 0.4

https://github.com/blakeblackshear/frigate/blob/8c7f6d4a7679dc42f49ea60291d24f69b9cf6e80/frigate/detectors/plugins/edgetpu_tfl.py#L72

and then after all detections are done, frigate reduces the detections before sending to the tracker

https://github.com/blakeblackshear/frigate/blob/8c7f6d4a7679dc42f49ea60291d24f69b9cf6e80/frigate/video.py#L700

https://github.com/blakeblackshear/frigate/blob/8c7f6d4a7679dc42f49ea60291d24f69b9cf6e80/frigate/util/object.py#L455

File: docs/docs/configuration/object_detectors.md (hunk 4 above)
Comment (MarcA711) [reply docs/docs/configuration/object_detectors.md:328]: Actually, the default is `None`. The image provides yolov8{n,s,m,l,x} models that can be chosen using the `yolov8_rknn_model`. 

This option allows the user to mount his own rknn model to the container and use it.

Comment (NickM-27) [reply docs/docs/configuration/object_detectors.md:328]: I think providing two separate mechanisms for users to select a different model is going to be confusing. I'd expect that we can make this work with the current mechanism

I would suggest that you could do something similar to TRT where the user can choose from one of the rknn models using an environment variable and that model will be downloaded before frigate starts up

Comment (MarcA711) [reply docs/docs/configuration/object_detectors.md:328]: Hm, but this way the user has to configure the yolov8 model using the docker-compose.yml. What if the user can pass something like "default-yolov8-nano", "default-yolov8-small", ... using the `model: path:` option if he wants to use one of the provided models and an actual path like "/config/model_cache/rknn/my-model.rknn" if he wants to use one model that he mounts to the container? This way everything can be done in the config using just the `model: path:` option.

Comment (NickM-27) [reply docs/docs/configuration/object_detectors.md:328]: that could work as well

Comment (MarcA711) [reply docs/docs/configuration/object_detectors.md:328]: Alright, I would actually prefer this way. Thank you for your feedback.

**Comment by MarcA711** [comment]
I will use the folder `/config/model_cache/rknn` to store downloaded models and also advise the user to place their own models there. But i have two questions:
- If there is a model located under `/config/model_cache/rknn/test_model.rknn`, should the user pass the full path to this file (like `/config/model_cache/rknn/test_model.rknn`) or just a path relative to the `model_cache` folder (like `rknn/test_model.rknn`) as `model: path:` option?
- I plan to still include a yolov8n model in the rk image, so that rknn detection works out of the box even for users without an internet connection on their server. Where should I place that file? Before, I used `/model/yolov8n-320x320.rknn`, is this okay?

**Comment by NickM-27** [comment]
1. full path should be passed same as for the tensorrt detector https://deploy-preview-6262--frigate-docs.netlify.app/configuration/object_detectors#configuration-parameters
2. I think it is fine as is 👍

**Comment by MarcA711** [comment]
Thank you very much.

Sorry to bother you again, but I have three more question about post processing of detections:
- Can I just return the top20 results, even if the confidence is smaller than 0.4? Or should I filter out results with confidence < 0.4 to stay consistent with the other detectors? From the code I can't see a point where it matters if the confidence is bigger than 0.4.
- Do the results need to be ordered? So that the detection with highest score comes first? Again, I wouldn't see why the order matters.
- Maybe we should pass the objects that the user wants to detect to the detector. It could be that the user wants to record only persons. The Cam captures a frame with two cars, two cats, some other objects and a person in the background. The person is recognized by the detector, but only has a low score of 0.5. The other stuff is also detected, but with a higher score. Since the post processor does no nms, one object can be multiple times inside the list. This way, it could be that the person will not be passed back to frigate and no recording will be made.

**Comment by NickM-27** [comment]
1. I think it would be best to remain consistent and filter on score 0.4.
2. The order matters in the return 20 logic since you want to make sure that the top 20 are returned in case there are more. After that it does not matter. 
3. I don't think that's necessary, frigate has separate regions for different tracked objects depending on where they are so getting 20 detections that push out what you want seems unlikely.

I think some of these may be good to address in the future but for now I think it is imperative that the detection implementations remain consistent and changes can be made across the board. We don't want assumptions downstream breaking due to inconsistencies

**Comment by MarcA711** [comment]
I guess I implemented all requested changes. If there is no more issue with the current PR, it could be merged after I test it a bit more (at least until tomorrow evening).

File: docs/docs/configuration/object_detectors.md [hunk 6]
```diff
@@ -325,4 +330,44 @@ model:                                # required
...
+Explanation for rknn specific options:
+- **core mask** controls which cores of your NPU should be used. This option applies only to SoCs with a multicore NPU (at the time of writing this in only the RK3588/S). The easiest way is to pass the value as a binary number. To do so, use the prefix `0b` and write a `0` to disable a core and a `1` to enable a core, whereas the last digit coresponds to core0, the second last to core1, etc. You also have to use the cores in ascending order (so you can't use core0 and core2; but you can use core0 and core1). Enabling more cores can reduce the inference speed, especially when using bigger models (see section below). Examples:
+  - `core_mask: 0b000` or just `core_mask: 0` let the NPU decide which cores should be used. Default and recommended value.
+  - `core_mask: 0b001` use only core0.
+  - `core_mask: 0b011` use core0 and core1.
+  - `core_mask: 0b110` use core1 and core2. **This does not** work, since core0 is disabled.
+
+### Choosing a model
+
+There are 5 yolov8 models that differ in size and therefore load the NPU more or less. In ascending order, with the top one being the smallest and least computationally intensive model:
```
Comment (NickM-27) [inline docs/docs/configuration/object_detectors.md:344]:
```suggestion
There are 5 default yolov8 models that differ in size and therefore load the NPU more or less. In ascending order, with the top one being the smallest and least computationally intensive model:
```

**Review by NickM-27 — CHANGES_REQUESTED** [review]

**Comment by NickM-27** [comment]
looks good to me besides a couple nits in the docs

**Review by NickM-27 — APPROVED** [review]

**Comment by MarcA711** [comment]
I thought about a change to this PR. Currently, the user needs to pass "default-yolov8n" to the "model: path:" option in order to use the yolov8n model that comes with the image. Maybe it is better to call it just "yolov8n", "yolov8m" etc. If the user wants to use his own model, he needs to pass an absolut path that begins with "/", so there shouldn't be any issues with that. What do you think is more user friendly, calling it "default-yolov8n", "default-yolov8m", ... or "yolov8n", "yolov8m", ...?

Moreover, there are two more changes, I would like to have merged before the next release:
- Apparently, other rockchip SoCs need specific libraries and models included in the image. So at the moment, only the rk3588/s is supported. Enabling the other SoCs has high priority and will be targeted next.
- I would also like to see go2rtc presets in the next release. I guess lots of users have h265 cams and want to transcode the recordings to h264 to allow playback in the browser etc. Where should I add docs about the go2rtc presets? As far as I can see, no other platform added go2rtc presets thus far. If there are only go2rtc presets for the rk platform, I could add it to the hardware acceleration docs. If other platform also implement those presets, maybe a new chapter "go2rtc presets" should be added below "FFmpeg presets".

**Comment by NickM-27** [comment]
1. I think it's fine as is, the documentation covers what is needed and regardless users should be copy and pasting the necessary default config 
2. Regarding go2rtc, no other platforms have go2rtc presets because there is no need, go2rtc supports them directly with `#hardware`. Putting the documentation in hardware acceleration makes sense to me

**Comment by MarcA711** [comment]
Alright. Regarding 2.: I will probably try to get these presets merged to go2rtc as well after I finished the integration here.

**Comment by NickM-27** [comment]
sounds good, just a heads up I don't believe we will want to update go2rtc to a major new version in 0.13

**Pull request merged**